File: Fall Week 7 Recursion/c++/templates/doubleplusgood.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <vector>
using namespace std;

// The numbers in the input
vector<string> nums; 
// Store all possible numbers in 'ans'; we provide template code for making them unique
vector<long long> ans;
void solve(int pos, string curnum, long long sum){
    // Reminder: to convert a string 's' to a long long int, use stoll(s)
    // We need long long ints because there can be 18 digits, but a long long int is large enough to store the answer
    // What base case goes here


    // Your recursive calls here
}
int main(){
    /* Reading and parsing the input*/
    string s;
    cin >> s;
    for(char& c : s) if(c == '+') c = ' ';
    stringstream ss(s);
    string tmp;
    while(ss >> tmp){
        nums.push_back(tmp);
    }
    /* End input*/

    // Begin recusion
    solve(0, "", 0); 

    // Find unique elements and print answer  
    sort(ans.begin(), ans.end());
    ans.erase(unique(ans.begin(), ans.end()), ans.end());
    cout << ans.size() << '\n';
}

File: Fall Week 7 Recursion/python/solutions/doubleplusgood.py
numbers = list(map(int, input().strip().split("+")))
n = len(numbers)
possible_outputs = set()


def generate_output(pos, curnum, summables):
    if pos == n:
        possible_outputs.add(curnum + sum(summables))
        return
    generate_output(pos+1, int(str(curnum) + str(numbers[pos])), summables)
    generate_output(pos+1, numbers[pos], summables + [curnum])


generate_output(0, 0, [])
print(len(possible_outputs))
